server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -42,32 +45,53 @@
 public class FieldFetcher {
     public static FieldFetcher create(QueryShardContext context,
                                       SearchLookup searchLookup,
-                                      Collection<FieldAndFormat> fieldAndFormats) {
+                                      Collection<FieldAndFormat> fieldAndFormats,
+                                      boolean includeUnmapped) {
 
         List<FieldContext> fieldContexts = new ArrayList<>();
+        List<String> originalPattern = new ArrayList<>();
+        List<String> excludeForUnmappedFetch = new ArrayList<>();
 
         for (FieldAndFormat fieldAndFormat : fieldAndFormats) {
             String fieldPattern = fieldAndFormat.field;
             String format = fieldAndFormat.format;
 
             Collection<String> concreteFields = context.simpleMatchToIndexNames(fieldPattern);
+            originalPattern.add(fieldAndFormat.field);
             for (String field : concreteFields) {
                 MappedFieldType ft = context.getFieldType(field);
                 if (ft == null || context.isMetadataField(field)) {
                     continue;
                 }
                 ValueFetcher valueFetcher = ft.valueFetcher(context, searchLookup, format);
+                excludeForUnmappedFetch.add(field);
                 fieldContexts.add(new FieldContext(field, valueFetcher));
             }
+            if (fieldPattern.charAt(fieldPattern.length() - 1) != '*') {
+                // not a prefix pattern, exclude potential sub-fields when fetching unmapped fields
+                excludeForUnmappedFetch.add(fieldPattern + ".*");
+            }
         }
+        Function<Map<String, ?>, Map<String, Object>> filter = XContentMapValues.filter(
```

Comment:
Re-using some source filtering logic seems to be a good fit. A couple ideas:
* Perhaps we shouldn't add the concrete field names as exclusions. The automata used for filtering have limits on the number of states, and I think we could run into these limits if the field patterns resolve to a large number of concrete fields. Here's an example exception from source filtering: https://github.com/elastic/elasticsearch/issues/53739.
* A slightly different approach would be to filter and collect the `DocumentField` entries at the same time, instead of doing two passes. This would involve writing a similar method to `XContentMapValues#filter`, but that flattens and collects the key-value pairs too (maybe something like `XContentMapValues#filterAndFlatten`?) I'm not sure this will turn out more cleanly, but wanted to mention the idea.

Reply:
Thanks for the feedback, I'll need to better understand at how XContentMapValues#filter works to see if and how this could work.